### PR TITLE
[Notifier] Introduce PHPUnit constraints and assertions for the Notifier

### PR DESCRIPTION
**Pull request by ismail1432**
| Q             | A
| ------------- | ---
| Branch?       | 6.2 
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no 
| Tickets       | Fix #... 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

[Since 4.4 we have assertions for email ](https://symfony.com/blog/new-in-symfony-4-4-phpunit-assertions-for-email-messages), the goal of this PR is to introduce the same kind of assertions for the Notifier component.

```php

   public function testNotifier()
      {  
         /** @var NotifierInterface $notifier */
         $firstNotification = new Notification('Hello World!', ['chat/slack']);
        $firstNotification->content("Symfony is awesome!");

        $notifier->send($firstNotification);

        $secondNotification = (new Notification('New urgent notification'))
            ->importance(Notification::IMPORTANCE_URGENT)
        ;
        $notifier->send($secondNotification);

                $this->assertNotificationCount(2);
        $first = 0;
        $second = 1;
        $this->assertNotificationIsNotQueued($this->getNotifierEvent($first));
        $this->assertNotificationIsNotQueued($this->getNotifierEvent($second));

        $notification = $this->getNotifierMessage($first);
        $this->assertNotificationSubjectContains($notification, 'Hello World!');
        $this->assertNotificationSubjectNotContains($notification, 'New urgent notification');
        $this->assertNotificationTransportIsEqual($notification,'slack');
        $this->assertNotificationTransportIsNotEqual($notification,'mercure');


        $notification = $this->getNotifierMessage($second);
        $this->assertNotificationSubjectContains($notification, 'New urgent notification');
        $this->assertNotificationSubjectNotContains($notification, 'Hello World!');
        $this->assertNotificationTransportIsEqual($notification,'mercure');
        $this->assertNotificationTransportIsNotEqual($notification,'slack');

  }

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Test/NotificationAssertionsTrait.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        return self::getMessageNotificationEvents()->getMessages($transport);
+    }
+
+    public static function getNotifierMessage(int $index = 0, string $transport = null): ?MessageInterface
+    {
+        return self::getNotifierMessages($transport)[$index] ?? null;
+    }
+
+    public static function getMessageNotificationEvents(): NotificationEvents
```
Comment (ismail1432) [inline src/Symfony/Bundle/FrameworkBundle/Test/NotificationAssertionsTrait.php]: getNotificationEvents

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Controller/NotifierController.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\TestBundle\Controller;
+
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Notifier\Notification\Notification;
+use Symfony\Component\Notifier\NotifierInterface;
+
+final class NotifierController
```
Comment (ismail1432) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Controller/NotifierController.php]: NotificationController

File: src/Symfony/Bundle/FrameworkBundle/Test/NotificationAssertionsTrait.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+/*
+ * @author Smaïne Milianni <smaine.milianni@gmail.com>
+ */
+use PHPUnit\Framework\Constraint\LogicalNot;
+use Symfony\Component\Notifier\Event\MessageEvent;
+use Symfony\Component\Notifier\Event\NotificationEvents;
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Test\Constraint as NotifierConstraint;
+
+trait NotificationAssertionsTrait
```
Comment (kbond) [inline src/Symfony/Bundle/FrameworkBundle/Test/NotificationAssertionsTrait.php]:
```suggestion
use PHPUnit\Framework\Constraint\LogicalNot;
use Symfony\Component\Notifier\Event\MessageEvent;
use Symfony\Component\Notifier\Event\NotificationEvents;
use Symfony\Component\Notifier\Message\MessageInterface;
use Symfony\Component\Notifier\Test\Constraint as NotifierConstraint;

/**
 * @author Smaïne Milianni <smaine.milianni@gmail.com>
 */
trait NotificationAssertionsTrait
```

**Review by kbond — COMMENTED** [review]
Nice work @ismail1432!

This makes sense to me and looks to match the email assertions. I'm not that familiar with notifier but should there be assertions on:

1. Who the notification was sent to?
2. What was in the body?

@OskarStark, you've been wanting this type of feature for some time and use notifier in your app(s). Do these assertions cover your needs?

**Comment by ismail1432** [comment]
> This makes sense to me and looks to match the email assertions. I'm not that familiar with notifier but should there be assertions on:
> 
> Who the notification was sent to?
> What was in the body?

I add more assertion, my point here is to introduce basic assertions and add more later by me or community :+1:

**Comment by kbond** [comment]
> my point here is to introduce basic assertions and add more later by me or community

Sure, that's fair!

File: src/Symfony/Component/Notifier/Message/NullMessage.php [hunk 4]
```diff
@@ -42,4 +44,13 @@ public function getTransport(): ?string
     {
         return $this->decoratedMessage->getTransport() ?? 'null';
     }
+
+    public function getNotification(): ?Notification
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Message/NullMessage.php]: Why is it needed?

Comment (ismail1432) [reply src/Symfony/Component/Notifier/Message/NullMessage.php]: AFAIK the notification is not exposed, so I added this method in order to access to the Notification’s properties and assert them in tests context.

File: src/Symfony/Component/Notifier/Test/Constraint/NotificationTransportIsEqual.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Test/Constraint/NotificationTransportIsEqual.php]: This must be removed.

File: src/Symfony/Component/Notifier/Test/Constraint/NotificationImportanceIsEqual.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Test/Constraint/NotificationImportanceIsEqual.php]: This must be removed.

File: src/Symfony/Component/Notifier/Message/NullMessage.php (hunk 4 above)
Comment (fabpot) [reply src/Symfony/Component/Notifier/Message/NullMessage.php]: But as `getNotification()` is not part of the interface, it's fragile. `EmailMessage` does not have a `getNotification()` method either. For now, I would remove constraints needing this.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Notifier/Message/NullMessage.php (hunk 4 above)
Comment (ismail1432) [reply src/Symfony/Component/Notifier/Message/NullMessage.php]: removed

File: src/Symfony/Bundle/FrameworkBundle/Test/NotificationAssertionsTrait.php [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        self::assertThat($event, new LogicalNot(new NotifierConstraint\NotificationIsQueued()), $message);
+    }
+
+    public static function assertNotificationSubjectContains(MessageInterface $messageObject, string $text, string $message = ''): void
+    {
+        self::assertThat($messageObject, new NotifierConstraint\NotificationSubjectContains($text), $message);
+    }
+
+    public static function assertNotificationSubjectNotContains(MessageInterface $messageObject, string $text, string $message = ''): void
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Test/NotificationAssertionsTrait.php:50]:
```suggestion
    public static function assertNotificationSubjectDoesNotContain(MessageInterface $messageObject, string $text, string $message = ''): void
```

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ismail1432.

**Pull request merged**